stdlib/public/core/KeyPath.swift

```diff
@@ -2501,10 +2658,39 @@ public func _swift_getKeyPath(pattern: UnsafeMutableRawPointer,
   let (keyPathClass, rootType, size, _)
     = _getKeyPathClassAndInstanceSizeFromPattern(patternPtr, arguments)
 
+  // Used to process offsets for pure struct KeyPaths.
+  var offset = UInt32(0)
+  var isPureStruct = true
+  var keyPathBase: Any.Type?
+        
   // Allocate the instance.
   let instance = keyPathClass._create(capacityInBytes: size) { instanceData in
     // Instantiate the pattern into the instance.
-    _instantiateKeyPathBuffer(patternPtr, instanceData, rootType, arguments)
+    let returnValue = _instantiateKeyPathBuffer(
+      patternPtr,
+      instanceData,
+      rootType,
+      arguments
+    )
+    offset += returnValue.structOffset
+
+    let componentStructList = returnValue.componentStructList
+    if componentStructList.count == 0 {
+      isPureStruct = false
+    }
+    for value in componentStructList {
+      isPureStruct = isPureStruct && value
+    }
+    keyPathBase = rootType
+  }
+
+  // TODO: See if there's a better way to eliminate
+  // tuples from the current `isPureStruct` optimization.
+  if let keyPathBase = keyPathBase, instance.isTuple(keyPathBase) {
+    isPureStruct = false
+  }
+  if isPureStruct {
```

Comment:
I don't think this check for a "pure struct" key path should be necessary. `_instantiateKeyPathBuffer` could return a nil offset if it encounters a key path component that isn't a stored struct component, and then you wouldn't assign the offset if there's no offset to assign.

Reply:
I've revised this as recommended.